getnacelle/nacelle-js
# feat: user-supplied custom GraphQL fragments

## Why are these changes introduced?

Addresses [ENG-6850](https://nacelle.atlassian.net/browse/ENG-6850)

> Allow users to supply custom fragments when initializing the client

Developers should have the freedom to query for additional Shopify cart fields so that they can accommodate their own use cases. This includes:
1. The ability to "extend" or make additive changes to the default GraphQL fragments used in `@nacelle/shopify-cart`.
2. The ability to totally replace some of the default GraphQL fragments.

## What is this pull request doing?

- Adds a `customFragments` parameter to the `CreateClientParams`.
- When `customFragments` are provided, they either:
  - Make additive changes to the default GraphQL fragments (this is the case for `EXTEND_CART`, `EXTEND_CART_LINE`, and `EXTEND_CART_LINE_MERCHANDISE`)
  - Totally replace default fragments (this is the case for `MONEY` and `USER_ERRORS`).
- When `customFragments` are provided, developers are free to use a fragment name of their choosing.
- Throws an error if `customFragments` are provided but are not in the expected format (e.g. an array).
- Throws an error if `customFragments` are provided but a fragment does not use the expected GraphQL type.

## How to test

1. Check out the `ENG-6850-dev-supplied-custom-fragments` branch.
2. `cd packages/shopify-cart`
3. `npm run test:ci` - all tests should pass with 100% coverage:

<img width="550" alt="ENG-6850-dev-supplied-custom-fragments coverage summary 100 percent coverage" src="https://user-images.githubusercontent.com/5732000/187768196-77f4ab01-2d79-4387-8ab2-e633bac51694.png">

## Next steps

These changes will be documented in a follow-up PR (for [ENG-6852](https://nacelle.atlassian.net/browse/ENG-6852)).

## End-to-end testing results

I spun up a Vite project and `npm link`ed to `@nacelle/shopify-cart`.

When providing `customFragments`, the data included the expected additional properties:

![shopify_cart_client_initialized_with_custom_fragments](https://user-images.githubusercontent.com/5732000/187769794-9d052745-a1e8-4f0a-81a2-c8baeea7cf1c.jpg)

![cart demo fragments sellingPlanAllocation and merchandise weight](https://user-images.githubusercontent.com/5732000/187769660-b4f68a36-d8c6-49e9-bac0-e7e04122f27c.png)

When I didn't provide `customFragments`, the cart data did not include any additional properties:

![cart demo without fragments](https://user-images.githubusercontent.com/5732000/187769644-2e5cc52e-f099-44d3-8ef5-ea7ed24c378f.png)
